### PR TITLE
Sample interface, new samplers,

### PR DESCRIPTION
**Pull request by ivanstepanovftw**
ignore EOS should apply -inf to EOS logit, new line penalization option, logit bias support (#1024)

New samplers:
- locally typical sampling
- tail free sampling
- frequency and presence penalty
- Mirostat & Mirostat v2

---

### Summary
🦙📝?🦙🧠?🦙🔧?

This pull request enhances the llama text generation library with new sampling techniques and features, such as logit bias, typicality filtering, frequency and presence penalties, mirostat, and newline penalty. It also updates the examples and the API to use the new sampling functions and structs, and to handle arrays of `llama_token_data`. It modifies the command line options and the usage message in `./examples/common.cpp` to reflect the new parameters and defaults.

> _We're coding with the llama, the llama of the sea_
> _We're sampling with the logits, the logits are the key_
> _We're adding new features, new features to the `gpt_params`_
> _We're heaving on the yardarm, on the yardarm, one, two, three_

### Walkthrough
*  Implement new sampling techniques and features for llama, such as tail free sampling, frequency and presence penalties, Mirostat sampling, logit bias, and newline penalty ([link](https://github.com/ggerganov/llama.cpp/pull/1126/files?diff=unified&w=0#diff-81037668d1ec4be4b72740f4070add30efb0b021c28e93e41c2a0a2062ba10e8R119-R130), [link](https://github.com/ggerganov/llama.cpp/pull/1126/files?diff=unified&w=0#diff-81037668d1ec4be4b72740f4070add30efb0b021c28e93e41c2a0a2062ba10e8R143-R172), [link](https://github.com/ggerganov/llama.cpp/pull/1126/files?diff=unified&w=0#diff-81037668d1ec4be4b72740f4070add30efb0b021c28e93e41c2a0a2062ba10e8L188-R253), [link](https://github.com/ggerganov/llama.cpp/pull/1126/files?diff=unified&w=0#diff-9ee6fe1df475323e1bc2457aa555d36474236452b27af58909fc410c9a5fe642L27-R39), [link](https://github.com/ggerganov/llama.cpp/pull/1126/files?diff=unified&w=0#diff-9ee6fe1df475323e1bc2457aa555d36474236452b27af58909fc410c9a5fe642L50-R59), [link](https://github.com/ggerganov/llama.cpp/pull/1126/files?diff=unified&w=0#diff-2d3599a9fad195f2c3c60bd06691bc1815325b3560b5feda41a91fa71194e805L405-R466), [link](https://github.com/ggerganov/llama.cpp/pull/1126/files?diff=unified&w=0#diff-a2f09a47e379eeeb66a7398e3a1d11a391af75829d3e7a6dd7218e221b4fcaf3L184-R234))
*  Update the command line options and parameters in the `common` files to reflect the new sampling techniques and features, and add descriptions and references for them in the usage message ([link](https://github.com/ggerganov/llama.cpp/pull/1126/files?diff=unified&w=0#diff-81037668d1ec4be4b72740f4070add30efb0b021c28e93e41c2a0a2062ba10e8R9-R10), [link](https://github.com/ggerganov/llama.cpp/pull/1126/files?diff=unified&w=0#diff-81037668d1ec4be4b72740f4070add30efb0b021c28e93e41c2a0a2062ba10e8L243-R327), [link](https://github.com/ggerganov/llama.cpp/pull/1126/files?diff=unified&w=0#diff-9ee6fe1df475323e1bc2457aa555d36474236452b27af58909fc410c9a5fe642R11), [link](https://github.com/ggerganov/llama.cpp/pull/1126/files?diff=unified&w=0#diff-9ee6fe1df475323e1bc2457aa555d36474236452b27af58909fc410c9a5fe642L20))
*  Update the message printed to the standard error stream in the `main` example to include the values of the new parameters ([link](https://github.com/ggerganov/llama.cpp/pull/1126/files?diff=unified&w=0#diff-2d3599a9fad195f2c3c60bd06691bc1815325b3560b5feda41a91fa71194e805L279-R280))
*  Modify the existing parameters and sampling logic in the `main` and `save-load-state` examples to use the new sampling techniques and features, and the new llama API functions ([link](https://github.com/ggerganov/llama.cpp/pull/1126/files?diff=unified&w=0#diff-2d3599a9fad195f2c3c60bd06691bc1815325b3560b5feda41a91fa71194e805L390-R402), [link](https://github.com/ggerganov/llama.cpp/pull/1126/files?diff=unified&w=0#diff-2a948c9cc9de5dc4f571997c720d4b1a61ff27ebf938f9e139fa063547eee768L67-R75), [link](https://github.com/ggerganov/llama.cpp/pull/1126/files?diff=unified&w=0#diff-2a948c9cc9de5dc4f571997c720d4b1a61ff27ebf938f9e139fa063547eee768L109-R118))
*  Modify the `llama_token_data` struct to store the logit instead of the plog, and add a new struct and a new function for handling arrays of `llama_token_data` ([link](https://github.com/ggerganov/llama.cpp/pull/1126/files?diff=unified&w=0#diff-a2f09a47e379eeeb66a7398e3a1d11a391af75829d3e7a6dd7218e221b4fcaf3L42-R51))

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -35,7 +35,7 @@ endif
 
 # keep standard at C11 and C++11
 CFLAGS   = -I.              -O3 -DNDEBUG -std=c11   -fPIC
-CXXFLAGS = -I. -I./examples -O3 -DNDEBUG -std=c++11 -fPIC
+CXXFLAGS = -I. -I./examples -O3 -DNDEBUG -std=c++20 -fPIC
```
Comment (Piezoid) [inline Makefile:38]: The last attempt to modernize the C++ standard was reverted (#355). I do agree that non-owning containers are nice, but it seems that the project wants to support old/underdeveloped compilers.

Comment (ivanstepanovftw) [reply Makefile:38]: But we are using C++ for internals, like examples, and expose only C API for both ggml and llama...

**Comment by Piezoid** [comment]
Nice work! 

I'll link the literature here. Feel free to complete with more up do date sources.
- [CTRL paper](https://arxiv.org/pdf/1909.05858.pdf) for the repetition penalty currently used in llama.cpp
- [Frequency and Presence penalties](https://platform.openai.com/docs/api-reference/parameter-details)
- [locally typical sampling](https://arxiv.org/pdf/2202.00666.pdf)
- [tail free sampling](https://www.trentonbricken.com/Tail-Free-Sampling/#tail-free-sampling-algorithm)
- [mirostat](https://arxiv.org/pdf/2007.14966.pdf)

I like the idea of a modular interface for sampling. It enables each code sample and application to combine these parts to do its own kitchen-sink sampling that fits their needs. Going further with this, the llama.h interface could be stripped to only provide access to logits and vocabulary, and the sampling code moved to a separate object file. This would emphasize and guarantee the extensibility of the samplers.

I am hesitant about the current implementation of repetition penalization.
As an illustration, I question whether the occurrence of past newlines and punctuation should guide the sampling of the following tokens. Attempting to fix this, the repetitions could be weighed against a simple frequency model. However, I wasn't able to recover such frequencies [from the tokenizer weights](https://github.com/ggerganov/llama.cpp/issues/331#issuecomment-1485167614).
It's possible to gather more information by measuring the length of the repetition that the next token would complete or interrupt. I have [implemented](https://github.com/Piezoid/llama.cpp/blob/0c49c52/llama.cpp#L1534-L1558) this idea and an exponential decay.

Concerning the application of the penalization, I'm not sure whether it is better to offset the logits or to scale them. Subtracting to the logit, used by "frequency and presence penalty", amounts to scaling the probabilities. Scaling the logits, which is discussed in the CTRL paper, can be thought of as a way of raising probabilities to a power, but is dependent on the logit=0 point which is not particularly meaningful.
Your current implementation applies both methods successively, which seems redundant. 

I haven't found the time to read in details about [mirostat](https://arxiv.org/pdf/2007.14966.pdf). My limited knowledge tells me that as the number of parameters goes up, the method becomes more challenging to apply in practice. Additionally, it seems difficult to control the changing target surprise `mu` using feedback, especially when working with an auto-regressive model. On the other hand, the promise of avoiding repetitions and boredom traps without looking at past tokens is very interesting.

I found that it is quite difficult to evaluate the sampling algorithms. We have good starting points with [your analysis](https://github.com/ggerganov/llama.cpp/discussions/997), the information-theoretic formalism of the [locally typical sampling](https://arxiv.org/pdf/2202.00666.pdf) and [mirostat](https://arxiv.org/pdf/2007.14966.pdf) papers, and their evaluation methods. Doing such experiments takes time end effort. Also, large scale human evaluations are next to impossible without a large community effort.

**Comment by ivanstepanovftw** [comment]
The CTRL paper does not mention, but in fact, the [CTRL repository](https://github.com/salesforce/ctrl/blob/a089a0c5956a84f229e5ece9d5da0bc4689dded0/pytorch_generation.py#L201-L208) explicitly avoids penalizing newline tokens during sampling.

File: Makefile (hunk 1 above)
Comment (Green-Sky) [reply Makefile:38]: what @Piezoid said

File: CMakeLists.txt [hunk 2]
```diff
@@ -75,7 +75,7 @@ option(LLAMA_BUILD_EXAMPLES         "llama: build examples" ${LLAMA_STANDALONE})
 # Compile flags
 #
 
-set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD 20)
```
Comment (Green-Sky) [inline CMakeLists.txt:78]: same as for make

**Review by Green-Sky — CHANGES_REQUESTED** [review]

File: tests/test-sampling.cpp [hunk 3]
```diff
@@ -0,0 +1,229 @@
+#include "ggml.h"
```
Comment (ggerganov) [inline tests/test-sampling.cpp:1]: `ggml.h` does not seem needed

**Review by ggerganov — APPROVED** [review]

**Comment by ivanstepanovftw** [comment]
Rebased, added 2 commits since last review

**Comment by ggerganov** [comment]
Mark "ready for review" when you think it is good for merge

**Comment by ivanstepanovftw** [comment]
I do not have Windows machine with MSVC installed, I am not sure why it fails:
```
3: Test command: D:\a\llama.cpp\llama.cpp\build\bin\Release\test-sampling.exe
3: Working Directory: D:/a/llama.cpp/llama.cpp/build/tests
3: Test timeout computed to be: 1500
3/4 Test #3: test-sampling ....................***Exception: Numerical  0.01 sec
```

**Comment by ivanstepanovftw** [comment]
Ready for review

**Review by Green-Sky — APPROVED** [review]
very cool. I always wanted a way to blacklist tokens, like backslash.

**Comment by ivanstepanovftw** [comment]
Oh, I got it, for `\code{begin}`!

**Comment by Green-Sky** [comment]
> Oh, I got it, for `\code{begin}`!

yea :smile: and `\code{end}` , the model often emits this before eos or tries do dodge/end the conversation.
Already tested it, works great.

**Comment by ivanstepanovftw** [comment]
You could write `-l 29905-inf` 😊
I have used stof instead of stringstream just to make `"inf"` work

**Comment by byroneverson** [comment]
Any thoughts on the removal of parameter defaults of new sampling function to keep llama.h compatible with C/Obj-C?

**Comment by DenisSergeevitch** [comment]
> edit: its `-l 29905-100000` , if anyone is interested.

Could anyone please share how to get the token id, and could I use multiple tokens at once with the --logit-bias flag?

**Comment by ivanstepanovftw** [comment]
> pass multiple tokens at once

Yes, by passing multiple arguments, like `./main ... -l 2-inf -l 13+2 -l 228+5`.

**Comment by DenisSergeevitch** [comment]
> > pass multiple tokens at once
> 
> Yes, by passing multiple arguments, like `./main ... -l 2-inf -l 13+2 -l 228+5`.

Thanks, I have done a small uncesoring [method](https://www.reddit.com/r/LocalLLaMA/comments/13j3747/tutorial_a_simple_way_to_get_rid_of_as_an_ai/) based on this flag, works like a charm!

**Comment by KerfuffleV2** [comment]
@ivanstepanovftw 
I'm working on a Rust-based implementation of these samplers and using the code you wrote as a reference. I'm crediting the llama.cpp project but I can mention by name in the project README as well since you wrote it (and I don't think it's really been changed much since the initial commit). I didn't want to do something like that without asking first, though.

Also, if you're unhappy with the way I'm handling this (the credits or otherwise) please let me know and hopefully we can work something out!

Link: https://github.com/KerfuffleV2/llm-samplers/

**Comment by ivanstepanovftw** [comment]
@KerfuffleV2 Sure you can! Glad that you support RWKV, looks very promising.
